src/Parser/DeathKnight/Unholy/Modules/Items/ColdHeartEfficiency.js

```diff
@@ -21,49 +21,93 @@ class ColdHeartEfficiency extends Analyzer {
     }
+	if (spellID === SPELLS.CONCORDANCE_OF_THE_LEGIONFALL_STRENGTH.id) {
+		this.concordanceStart = event.timestamp;
+	}
+	if (spellID === SPELLS.KHAZGOROTHS_SHAPING.id) {
+		this.khazgorothStart = event.timestamp;
+	}
   }
-
+  
+  on_byPlayer_refreshbuff(event) {
+	const spellID = event.ability.guid;
+	if (spellID === SPELLS.UNHOLY_STRENGTH_BUFF.id) {
+      this.unholyStrengthStart = event.timestamp;
+    }
+	if (spellID === SPELLS.CONCORDANCE_OF_THE_LEGIONFALL_STRENGTH.id) {
+		this.concordanceStart = event.timestamp;
+	}
+	if (spellID === SPELLS.KHAZGOROTHS_SHAPING.id) {
+		this.khazgorothStart = event.timestamp;
+	}
+  }
+  
   on_byPlayer_removebuff(event) {
     const spellID = event.ability.guid;
     if (spellID === SPELLS.UNHOLY_STRENGTH_BUFF.id) {
       this.unholyStrengthStart = 0;
     }
+	if (spellID === SPELLS.CONCORDANCE_OF_THE_LEGIONFALL_STRENGTH.id) {
+		this.concordanceStart = 0;
+    }
+	if (spellID === SPELLS.KHAZGOROTHS_SHAPING.id) {
+		this.khazgorothStart = 0;
+	}
   }
 
   on_byPlayer_applybuffstack(event) {
     const spellID = event.ability.guid;
     if (spellID === SPELLS.COLD_HEART_BUFF.id) {
       this.buffColdHeart = event.stack;
+	  if (this.buffColdHeart === 20) {
+		  this.timeAtMaxStacksStart = event.timestamp;
+	  }
     }
   }
 
   on_byPlayer_cast(event) {
     const spellId = event.ability.guid;
+	
     const unholyStrengthDuration = 15000;
+	const concordanceDuration = 10000;
+	const khazgorothDuration = 15000;
+	
     const coldHeartMaxStack = 20;
-    const coldHeartOptimalMinStack = 17;
-    const remainingUnholyStrengthAllowed = 4000;
+    const coldHeartOptimalMinStack = 14;
+	
+    const remainingDurationAllowed = 4000;
+	
+	this.timeAtMaxStacks = event.timestamp - this.timeAtMaxStacksStart;
+	
     this.unholyStrengthRemaining = unholyStrengthDuration - (event.timestamp - this.unholyStrengthStart);
+	this.concordanceRemaining = concordanceDuration - (event.timestamp - this.concordanceStart);
+	this.khazgorothRemaining = khazgorothDuration - (event.timestamp - this.khazgorothStart);
+	
     if (spellId === SPELLS.CHAINS_OF_ICE.id) {
       this.totalColdHeartCasts++;
-      if (this.unholyStrengthRemaining > 0) {
-        if (this.unholyStrengthRemaining < remainingUnholyStrengthAllowed) {
-          if (this.buffColdHeart < coldHeartMaxStack) {
-            if (this.buffColdHeart >= coldHeartOptimalMinStack) {
-              this.correctColdHeartCasts++;
-            }
-          }
-        }
+	  //This checks wether or not any of the three buffs are about to fall off within the next 4 seconds.
+      if ((this.unholyStrengthRemaining > 0 & this.unholyStrengthRemaining < remainingDurationAllowed) || (this.concordanceRemaining > 0 & this.concordanceRemaining < remainingDurationAllowed) || (this.khazgorothRemaining > 0 & this.khazgorothRemaining < remainingDurationAllowed) & this.buffColdHeart < coldHeartMaxStack & this.buffColdHeart >= coldHeartOptimalMinStack) {
```

Comment:
Couple of notes, first make sure to use `&&`.  `&` is a bitwise operator, `&&` is the logical operator.  This should also be broken into two if statements to make it more readable.  Something `if (this.buffColdHeart < coldHeartMaxStack & this.buffColdHeart >= coldHeartOptimalMinStack)` on one line and then under that do the rest.  It'll make it much easier to see whats going at a glance.